src/fileio/impls/hdf5/h5_format.cxx

```diff
@@ -253,6 +253,146 @@ bool H5Format::setRecord(int t) {
+      throw BoutException("Failed to create attribute_dataspace");
+
+    // Create new string datatype for attribute
+    hid_t variable_length_string_type = H5Tcopy(H5T_C_S1);
+    if (variable_length_string_type < 0)
+      throw BoutException("Failed to create variable_length_string_type");
+    if (H5Tset_size(variable_length_string_type, H5T_VARIABLE) < 0)
+      throw BoutException("Failed to create string type");
+
+    // Create attribute and write to it
+    hid_t myatt_in = H5Acreate(dataSet, "bout_type", variable_length_string_type, attribute_dataspace, H5P_DEFAULT, H5P_DEFAULT);
+    if (myatt_in < 0)
+      throw BoutException("Failed to create attribute");
+    if (H5Awrite(myatt_in, variable_length_string_type, &datatype) < 0)
+      throw BoutException("Failed to write attribute");
+
+    if (H5Pclose(propertyList) < 0)
+      throw BoutException("Failed to close propertyList");
+    if (H5Sclose(init_space) < 0)
+      throw BoutException("Failed to close init_space");
+    if (H5Sclose(attribute_dataspace) < 0)
+      throw BoutException("Failed to close attribute_dataspace");
+    if (H5Tclose(variable_length_string_type) < 0)
+      throw BoutException("Failed to close variable_length_string_type");
+    if (H5Aclose(myatt_in) < 0)
+      throw BoutException("Failed to close myatt_in");
+  } else {
+    dataSet = H5Dopen(dataFile, name.c_str(), H5P_DEFAULT);
+    if (dataSet < 0) {
+      // Negative value indicates error, i.e. file does not exist, so create:
+      hsize_t init_size[3],init_size_local[3];
+      if (parallel) {
+        init_size[0]=mesh->GlobalNx-2*mesh->xstart; init_size[1]=mesh->GlobalNy-2*mesh->ystart; init_size[2]=mesh->GlobalNz;
+      }
+      else {
+        init_size[0]=mesh->LocalNx; init_size[1]=mesh->LocalNy; init_size[2]=mesh->LocalNz;
+      }
+
+      // Create value for attribute to say what kind of field this is
+      std::string datatype = "scalar";
+      if(nd > 0) datatype = "FieldX";
+      if(nd == 2) datatype = "Field2D";
+      if(nd == 3) datatype = "Field3D";
+
+      if (nd==0) {
+        // Need to write a scalar, not a 0-d array
+        nd = 1;
+        init_size[0] = 1;
+      }
+
+      hid_t init_space = H5Screate_simple(nd, init_size, init_size);
+      if (init_space < 0)
+        throw BoutException("Failed to create init_space");
+      dataSet = H5Dcreate(dataFile, name.c_str(), write_hdf5_type, init_space, H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
+      if (dataSet < 0)
+        throw BoutException("Failed to create dataSet");
+
+      // Add attribute to say what kind of field this is
+      setAttribute(dataSet, "bout_type", datatype);
+    }
+  }
+
+  if (H5Dclose(dataSet) < 0)
+    throw BoutException("Failed to close dataSet");
+  return true;
+}
+
+bool H5Format::addVar_int(const string &name, bool repeat) {
+  return addVar(name, repeat, H5T_NATIVE_INT, 0);
+}
+
+bool H5Format::addVar_BoutReal(const string &name, bool repeat) {
+  if(lowPrecision) {
```

Comment:
To cut down on repeated code, this could maybe be written as:

```cpp
auto h5_float_type = lowPrecision ? H5T_NATIVE_FLOAT : H5T_NATIVE_DOUBLE;
return addVar(name, repeat, h5_float_type, 0);
```

It looks like using this pattern might help a lot here